Frameworks/MQTTnet.NetStandard/Serializer/MqttPacketReader.cs

```diff
@@ -80,13 +81,15 @@ public byte[] ReadRemainingData()
             return ReadBytes(_header.BodyLength - (int)BaseStream.Position);
         }
 
-        private static int ReadBodyLengthFromSource(Stream stream, CancellationToken cancellationToken)
+        private static async Task<int> ReadBodyLengthFromSourceAsync(Stream stream, CancellationToken cancellationToken)
```

Comment:
Did you test this change with the performance test app? I agree that the readByte method should be changed because it is not async but reading the body length should not be a problem. If a previous header was received (1 byte) there should be more data and thus there is no need to use async because reading single bytes all the time async will decrease the performance a lot. But as I said the other place with the header makes sense because the thread is being blocked (instead of freed) until the next message is received.

Reply:
I did not yet test the performance. Note however, that in TCP/IP, data can be split at any location (as TCP is steam-like), so even if the first byte arrived (for a new packet), it does not mean that the second, third etc. bytes are also already available. For example, a slow client might send the first byte, then wait some time before sending the remaining bytes; or a interceptor might split the bytes at this boundary etc.

Therefore, if the first byte is read with an async method but the remaining length bytes are read using a sync method, there is still a possibility that worker threads are blocked at this location if many clients do this.

Alternatively, I think we could use [`NetworkStream.DataAvailable`](https://docs.microsoft.com/de-de/dotnet/api/system.net.sockets.networkstream.dataavailable?view=netframework-4.7#System_Net_Sockets_NetworkStream_DataAvailable) before either using an async read method (if no data available) or a synchronous read method (if data is available), if using the async read method only has a negative performance impact.